File: 118. FindFirstAndLastPosition.c
#include <stdio.h>

static int findBound(int *nums, int n, int target, int isFirst) {
    int left = 0;
    int right = n - 1;
    int answer = -1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (nums[mid] == target) {
            answer = mid;
            if (isFirst) right = mid - 1;
            else left = mid + 1;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return answer;
}

static void searchRange(int *nums, int n, int target) {
    if (n == 0) {
        printf("[-1, -1]\n");
        return;
    }
    int first = findBound(nums, n, target, 1);
    if (first == -1) {
        printf("[-1, -1]\n");
        return;
    }
    int last = findBound(nums, n, target, 0);
    printf("[%d, %d]\n", first, last);
}

int main(void) {
    int nums1[] = {5, 7, 7, 8, 8, 10};
    searchRange(nums1, 6, 8);
    searchRange(nums1, 6, 6);
    searchRange(NULL, 0, 0);
    return 0;
}

File: 118. FindFirstAndLastPosition.java
import java.util.Arrays;

public class findFirstAndLastPosition {

    public static int[] searchRange(int[] nums, int target) {
        int first = findBound(nums, target, true);
        if (first == -1) {
            return new int[] {-1, -1};
        }

        int last = findBound(nums, target, false);
        return new int[] {first, last};
    }

    private static int findBound(int[] nums, int target, boolean isFirst) {
        int left = 0;
        int right = nums.length - 1;
        int answer = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) {
                answer = mid;
                if (isFirst) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return answer;
    }

    public static void main(String[] args) {
        System.out.println(Arrays.toString(searchRange(new int[] {5, 7, 7, 8, 8, 10}, 8)));
        System.out.println(Arrays.toString(searchRange(new int[] {5, 7, 7, 8, 8, 10}, 6)));
        System.out.println(Arrays.toString(searchRange(new int[] {}, 0)));
    }
}
